hw/mcu/stm/stm32l1xx/src/hal_flash.c

```diff
@@ -65,31 +65,44 @@ static int
 stm32l1_flash_write(const struct hal_flash *dev, uint32_t address,
         const void *src, uint32_t num_bytes)
 {
-    const uint32_t *sptr;
+    uint32_t val;
     uint32_t i;
     int rc;
+    uint8_t align;
     uint32_t num_words;
 
     if (!num_bytes) {
         return -1;
     }
 
+    align = dev->hf_align;
+    num_words = ((num_bytes - 1) / dev->hf_align) + 1;
```

Comment:
Divide is expensive, bitwise operations are cheap. I would use those rather, given that alignment is a power of 2.

Reply:
Yup, agreed but `udiv` is not thaaaat bad. I guess it will be worse on L0 which is Cortex-M0 and will need emulation. Btw, I took a look at the generated listing and there are only 3 divisions, this one + two in the clock initialization code inside the HAL!

Reply:
Btw, I don't know of any bit-fiddling that can do `log2` so I would use a small lookup table (alignments are 1 to 8 only) or simply a `switch`, but I wonder how better that would be in number of cycles than just doing a `udiv` here. It's probably much better on the L0 but may be worse on the L1/L4!

Reply:
Luckily I have this code around. And actually, I'd be quite interested in seeing what different it does make :)
https://github.com/mkiiskila/incubator-mynewt-core/tree/flash_speed_test

Include test/flash_test package with slinky, and then from console:
'flash_speed 0 0x08000000 0 0 1'
I admit, the arguments to this might not be the greatest.

If you have multiple possible alignments, and this code will be shared by all, I'd do:
#if FLASH_ALIGNMENT == 8
#define FLASH_ALIGNMENT_SHIFT 3
#elif FLASH_ALIGNMENT == 4
#define FLASH_ALIGNMENT_SHIFT 2
...

num_words = (((num_bytes - 1) >> FLASH_ALIGNMENT_SHIFT) + 1;

Anyway, running the speed test would show if udiv is worth avoiding or not.